core/src/main/java/io/aiven/kafka/tieredstorage/metadata/SegmentCustomMetadataBuilder.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2023 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tieredstorage.metadata;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+import org.apache.kafka.server.log.remote.storage.RemoteLogSegmentMetadata;
+
+import io.aiven.kafka.tieredstorage.ObjectKey;
+
+public class SegmentCustomMetadataBuilder {
+    final ObjectKey objectKey;
+    final RemoteLogSegmentMetadata segmentMetadata;
+    final Map<ObjectKey.Suffix, UploadResult> uploadResults;
+
+    final Set<SegmentCustomMetadataField> fields;
+
+    public SegmentCustomMetadataBuilder(final Set<SegmentCustomMetadataField> fields,
+                                        final ObjectKey objectKey,
+                                        final RemoteLogSegmentMetadata segmentMetadata) {
+        this.fields = fields;
+        this.objectKey = objectKey;
+        this.segmentMetadata = segmentMetadata;
+        this.uploadResults = new HashMap<>(ObjectKey.Suffix.values().length);
+    }
+
+    public SegmentCustomMetadataBuilder addUploadResult(final ObjectKey.Suffix suffix,
+                                                        final String keyPath,
+                                                        final long bytes) {
+        if (uploadResults.containsKey(suffix)) {
+            throw new IllegalArgumentException("Upload results for suffix " + suffix + " already added");
+        }
+        uploadResults.put(suffix, new UploadResult(keyPath, bytes));
+        return this;
+    }
+
+    public long totalSize() {
+        return uploadResults.values().stream().mapToLong(value -> value.bytes).sum();
+    }
+
+    public Map<Integer, Object> build() {
```

Comment:
Tagged fields [require](https://github.com/apache/kafka/blob/f44ee4fab7ef7adf715ecf2b96defa5cc8311949/clients/src/main/java/org/apache/kafka/common/protocol/types/TaggedFields.java#L153) `NavigableMap`, we should use it as the type instead of just `Map`. I know we pass `TreeMap` anyway, but it makes sense to ensure this on the type level.

Reply:
Agree. Applying suggestion.